packages/hash/api/src/storage/index.ts

```diff
@@ -1,3 +1,146 @@
+
+// S3-like APIs have a 7 day upper bound.
+const DOWNLOAD_URL_EXPIRATION_SECONDS = 60 * 60 * 24 * 7;
+// An offset for the cached URL to prevent serving invalid URL
+const DOWNLOAD_URL_CACHE_OFFSET = 60 * 60 * 1;
+
+/** Helper type to create a typed "dictionary" of storage types to their storage provider instance */
+export type StorageProviderLookup = Partial<
+  Record<StorageType, StorageProvider | UploadableStorageProvider>
+>;
+
+type StorageProviderInitialiser = (
+  app: Express,
+) => StorageProvider | UploadableStorageProvider;
+
+const storageProviderInitialiserLookup: Record<
+  StorageType,
+  StorageProviderInitialiser
+> = {
+  [StorageType.AwsS3]: (_app: Express) =>
+    new AwsS3StorageProvider(getAwsS3Config()),
+  [StorageType.ExternalLink]: (_app: Express) => new ExternalStorageProvider(),
+  [StorageType.LocalFileSystem]: (app: Express) =>
+    new LocalFileSystemStorageProvider({
+      app,
+      fileUploadPath: LOCAL_FILE_UPLOAD_PATH,
+      apiOrigin,
+    }),
+};
+/** All storage providers usable by the API should be added here.
+ * Even if not currently used for upload, they need to be available for downloads.
+ */
+const storageProviderLookup: StorageProviderLookup = {};
+let uploadStorageProvider: StorageType = StorageType.LocalFileSystem;
+
+const initialiseStorageProvider = (app: Express, provider: StorageType) => {
+  const initialiser = storageProviderInitialiserLookup[provider];
+
+  const newProvider = initialiser(app);
+  storageProviderLookup[provider] = newProvider;
+  return newProvider;
+};
+
+export const getStorageProvider = (
+  app: Express,
+  provider: StorageType,
+): StorageProvider => {
+  if (storageProviderLookup[provider]) {
+    return storageProviderLookup[provider]!;
+  } else {
+    return initialiseStorageProvider(app, provider);
+  }
+};
+
+export const getUploadStorageProvider = (): UploadableStorageProvider => {
+  const uploadProvider = storageProviderLookup[uploadStorageProvider];
+  if (!uploadProvider) {
+    throw new Error(
+      `Upload storage provider ${uploadStorageProvider} is required by the app but doesn't exist`,
+    );
+  }
+  return uploadProvider as UploadableStorageProvider;
+};
+
+export const setupStorageProviders = (
+  app: Express,
+  fileUploadProvider: StorageType,
+): UploadableStorageProvider => {
+  initialiseStorageProvider(app, fileUploadProvider);
+  uploadStorageProvider = fileUploadProvider;
+  return getUploadStorageProvider();
+};
+
+/**
+ * Set up express route to proxy uploaded files so we can cache presigned URLs.
+ *
+ * @todo We should consider authorization for this route (it doesn't even authenticate for now.)
+ *   https://app.asana.com/0/1200211978612931/1202510174412958/f
+ *
+ * @param app - the express app
+ * @param storageProvider - the provider we're using for file storage
+ * @param cache - a cache to store presigned URLs so we don't needlessly create URLs for every download
+ */
+export const setupFileProxyHanldere = (
```

Comment:
Nit: typo
```suggestion
export const setupFileProxyHandler = (
```

Reply:
Shoot, missed this one: d2a4490393a6a39f5b9210f43beef8a3b28e3a4f